### PR TITLE
A new gate_sequence_product, Simulator and Results Class

### PR DESCRIPTION
**Pull request by sarsid**
**Checklist**
Thank you for contributing to QuTiP! Please make sure you have finished the following tasks before opening the PR.

- [x] Please read [Contributing to QuTiP Development](https://github.com/qutip/qutip-doc/blob/master/CONTRIBUTING.md)
- [x] Contributions to qutip should follow the [pep8 style](https://www.python.org/dev/peps/pep-0008/).
You can use [pycodestyle](http://pycodestyle.pycqa.org/en/latest/index.html) to check your code automatically
- [x] Please add tests to cover your changes if applicable.
- [x] If the behavior of the code has changed or new feature has been added, please also update the [documentation](https://github.com/qutip/qutip-doc) and the [notebook](https://github.com/qutip/qutip-notebooks). Feel free to ask if you are not sure.

**Description**
There are three major parts to this PR (more or less interlinked): 

- Adds `Simulator` class to abstract away circuit run functionality. This enables future contributors to add to exact simulation (and potentially other ways of simulation) easily. It also has a `step` function which allows to go over the circuit step-by-step which could enable interesting visualisations.
- Adds `Result` class as a wrapper around results obtained from circuit simulations. 
- Adds a more efficient mode to `gate_sequence_product` function which allows for faster pre-computation of unitaries.

**Changelog**
Added `Simulator`, `Result` class 
Added new mode to `gate_sequence_product`, `QubitCircuit.propagators`

### CODE REVIEW

**Comment by BoxiLi** [comment]
@sarsid, could you push the code you have on this PR? I'd like to get it merged and use the Result class you defined also in the Processor class ;)

File: qutip/qip/operations/gates.py [hunk 1]
```diff
@@ -931,24 +931,158 @@ def hadamard_transform(N=1):
...
+    U_sublist = U * U_sublist
+    inds_sublist = revised_inds
+
+    overall_inds_revised.append(inds_sublist)
+    tensor_list_revised.append(U_sublist)
+
+    return tensor_list_revised, overall_inds_revised
+
+
+def _gate_sequence_product(U_list, ind_list, expand_N=None):
```
[the remark this answers is not in the document]
Comment (sarsid) [reply qutip/qip/operations/gates.py:1006]: Yep, thanks for spotting this, do not need this parameter anymore

File: qutip/qip/operations/gates.py [hunk 2]
```diff
@@ -959,6 +1093,41 @@ def gate_sequence_product(U_list, left_to_right=True):
...
+    -------
+    U_overall : qobj
+        Unitary matrix corresponding to U_list.
+
+    overall_inds : list of int, optional
+        List of qubit indices on which U_overall applies.
+    """
+
+    if expand:
+        return _gate_sequence_product(U_list, inds_list)
```
[the remark this answers is not in the document]
Comment (sarsid) [reply qutip/qip/operations/gates.py:1126]: I have a few comments on this in the documentation.

File: qutip/qip/circuit.py [hunk 3]
```diff
@@ -1710,3 +1751,459 @@ def _to_qasm(self, qasm_out):
...
+        qc: :class:`.QubitCircuit`
+            Quantum Circuit to be simulated.
+
+        state: ket or oper
+            ket or density matrix
+
+        cbits: list of int, optional
+            initial value of classical bits
+
+        U_list: list of Qobj, optional
```
[the remark this answers is not in the document]
Comment (sarsid) [reply qutip/qip/circuit.py:1861]: I kept *U_list* argument in the optional case where a user calculated it for some reason before hand. Should we remove it? Maybe not needed anymore. Don't think it goes well with the other options anymore.

File: qutip/qip/operations/gates.py [hunk 4]
```diff
@@ -931,24 +931,158 @@ def hadamard_transform(N=1):
...
+        else:
+            # only need to expand stuff !
+            overall_inds.append(inds)
+            tensor_list.append(U)
+
+    U_overall = tensor(tensor_list)
+    return U_overall, [sorted_inds[ind] for ind in _flatten(overall_inds)]
+
+
+def _gate_sequence_product_expanded(U_list, left_to_right=True):
```
[the remark this answers is not in the document]
Comment (sarsid) [reply qutip/qip/operations/gates.py:1068]: Agreed

**Comment by sarsid** [comment]
@BoxiLi should be good to go, added more inline docs, plus some examples in `gates.py`

File: qutip/qip/operations/gates.py [hunk 5]
```diff
@@ -1016,56 +1057,80 @@ def _gate_sequence_product(U_list, ind_list, expand_N=None):
...
+        # tensor_list
         else:
-            # only need to expand stuff !
             overall_inds.append(inds)
             tensor_list.append(U)
 
-    U_overall = tensor(tensor_list)
-    return U_overall, [sorted_inds[ind] for ind in _flatten(overall_inds)]
+    U_overall, overall_inds = _expand_overall(tensor_list, overall_inds)
+    print(overall_inds)
```
Comment (BoxiLi) [inline qutip/qip/operations/gates.py:1129]: I guess this is only for local test

**Review by BoxiLi — CHANGES_REQUESTED** [review]
Thanks @sarsid . Looks good to me, except for a remaining `print`. With this removed, it is ready for merging.

File: qutip/qip/operations/gates.py (hunk 5 above)
Comment (sarsid) [reply qutip/qip/operations/gates.py]: done

**Review by BoxiLi — APPROVED** [review]

**Comment by nathanshammah** [comment]
Well done @sarsid. Could you make sure that the documentation is still up to date? Adding something in the Users Guide about this would be very helpful.

**Comment by sarsid** [comment]
Thanks @nathanshammah, I believe that most of the docs are up to date in the PR https://github.com/qutip/qutip-doc/pull/117, I will make sure to make a couple of changes Boxi suggested before it's ready to merge.

**Pull request merged**